x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/ResultsIndexUpgradeService.java

```diff
@@ -0,0 +1,523 @@
+        oldAliases.forEach((oldIndex, aliases) ->
+            {
+                if (aliases.isEmpty() == false) { //if the aliases are empty, that means there are none to remove
+                    indicesAliasesRequest.addAliasAction(IndicesAliasesRequest
+                        .AliasActions
+                        .remove()
+                        .index(oldIndex)
+                        .aliases(aliases.stream().map(Alias::name).toArray(String[]::new)));
+                }
+            }
+        );
+        newAliases.forEach((newIndex, aliases) ->
+            aliases.forEach(alias -> {
+                IndicesAliasesRequest.AliasActions action = IndicesAliasesRequest.AliasActions.add().index(newIndex);
+                if (alias.filter() != null) {
+                    action.filter(alias.filter());
+                }
+                action.alias(alias.name());
+                indicesAliasesRequest.addAliasAction(action);
+            })
+        );
+        executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+            ML_ORIGIN,
+            indicesAliasesRequest,
+            indicesAliasListener,
+            client.admin().indices()::aliases);
+    }
+
+
+    private static class IndexNameAndAliasProvider {
+
+        private final List<String> oldIndices;
+        private final Map<String, List<Alias>> writeAliases = new HashMap<>();
+        private final Map<String, List<Alias>> readAliases = new HashMap<>();
+
+        private IndexNameAndAliasProvider(List<String> oldIndices, MetaData metaData) {
+            this.oldIndices = oldIndices;
+            oldIndices.forEach(index -> {
+                IndexMetaData indexMetaData = metaData.index(index);
+                List<Alias> writes = new ArrayList<>();
+                List<Alias> reads = new ArrayList<>();
+                indexMetaData.getAliases().forEach(aliasCursor -> {
+                    Alias alias = new Alias(aliasCursor.value.alias());
+                    if (aliasCursor.value.filteringRequired()) {
+                        alias.filter(aliasCursor.value.getFilter().string()); //Set the read alias jobId filter
+                    }
+                    if (alias.name().contains(".write-")) {
+                        writes.add(alias);
+                    } else {
+                        reads.add(alias);
+                    }
+                });
+
+                writeAliases.put(index, writes);
+                readAliases.put(index, reads);
+            });
+        }
+
+        private Exception validate(MetaData metaData, Predicate<IndexMetaData> shouldUpgrade) {
+            for (String index : oldIndices) {
+                IndexMetaData indexMetaData = metaData.index(index);
+
+                if (indexMetaData.getState() != IndexMetaData.State.OPEN) {
+                    return new IllegalStateException("unable to upgrade a closed index[" + index + "]");
```

Comment:
Does this ever happen given that you're forbidding closed indices when expanding wildcards?

Reply:
No, this doesn't happen. I can remove this check. It was more of an insurance if the indicesOption for the action request did not error out on closed indices.